snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/SnomedConceptRequestCache.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2021 B2i Healthcare Pte Ltd, http://b2i.sg
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.b2international.snowowl.snomed.datastore.request;
+
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import com.b2international.commons.http.ExtendedLocale;
+import com.b2international.commons.options.Options;
+import com.b2international.snowowl.core.domain.BranchContext;
+import com.b2international.snowowl.core.domain.IComponent;
+import com.b2international.snowowl.snomed.core.domain.SnomedConcept;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedSet;
+
+/**
+ * @since 8.1.0
+ */
+public class SnomedConceptRequestCache {
+
+	private final Map<FetchConfig, Map<String, SnomedConcept>> cache = new HashMap<>(); 
+	private final Deque<FetchConfig> requestedFetches = new ArrayDeque<>();
+	
+	public void request(Iterable<String> ids, Options expand, List<ExtendedLocale> locales, Consumer<Map<String, SnomedConcept>> onConceptsReady) {
+		requestedFetches.add(new FetchConfig(ids, expand, locales, onConceptsReady));
+	}
+	
+	public Map<String, SnomedConcept> get(BranchContext context, Iterable<String> ids, Options expand, List<ExtendedLocale> locales) {
+		FetchConfig config = new FetchConfig(ids, expand, locales, concepts -> {});
+		if (cache.containsKey(config)) {
+			return cache.get(config);
+		} else {
+			requestedFetches.add(config);
+			compute(context);
```

Comment:
Would it make sense to restrict computation to this config or configs with the same `expand` and `locale` parameters only, and not compute all pending requests?

Reply:
That was the initial implementation 😄 I can make an alternative, where there is a limit to the computed configs.

Reply:
Done